src/radical/pilot/pmgr/launching/base.py

```diff
@@ -25,34 +118,995 @@ def __init__(self, cfg, session):
+        self._active_sds = dict()
+        self._sds_lock   = mt.Lock()
+
+        self._log.info(ru.get_version([self._mod_dir, self._root_dir]))
+        self._rp_version, _, _, _, self._rp_sdist_name, self._rp_sdist_path = \
+                ru.get_version([self._mod_dir, self._root_dir])
+
+
+        # load all launcher implementations
+        self._launchers = dict()
+
+        from .saga  import PilotLauncherSAGA
+        from .psi_j import PilotLauncherPSIJ
+
+        impl = {
+            RP_UL_NAME_SAGA : PilotLauncherSAGA,
+            RP_UL_NAME_PSI_J: PilotLauncherPSIJ
+        }
+
+        for name in [RP_UL_NAME_PSI_J, RP_UL_NAME_SAGA]:
+            try:
+                ctor = impl[name]
+                self._launchers[name] = ctor(name, self._log, self._prof,
+                                             self._state_cb)
+            except:
+                self._log.exception('skip launcher %s' % name)
 
 
     # --------------------------------------------------------------------------
     #
-    # This class-method creates the appropriate sub-class for the Launcher.
-    #
     @classmethod
     def create(cls, cfg, session):
 
-        name = cfg.get('pmgr_launching_component', RP_UL_NAME_DEFAULT)
+        return cls(cfg, session)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _state_cb(self, pilot, rp_state):
 
-        # Make sure that we are the base-class!
-        if cls != PMGRLaunchingComponent:
-            raise TypeError("Factory only available to base class!")
+        self._log.info('pilot state update: %s: %s', pilot['uid'], rp_state)
 
-        from .default import Default
+        if rp_state != pilot['state']:
+            self.advance(pilot, state=rp_state, push=False, publish=True)
 
-        impl = {
-            RP_UL_NAME_DEFAULT: Default
-        }
 
-        if name not in impl:
-            raise ValueError('PMGRLaunchingComponent %s unknown' % name)
+    # --------------------------------------------------------------------------
+    #
+    def finalize(self):
+
+        # FIXME: is this called?
+
+        try:
+            self.unregister_input(rps.PMGR_LAUNCHING_PENDING,
+                                  rpc.PMGR_LAUNCHING_QUEUE, self.work)
+
+            # FIXME: always kill all pilot jobs for non-final pilots at
+            #        termination, and set the pilot states to CANCELED.
+            with self._lock:
+                pids = list(self._pilots.keys())
+
+            self._kill_pilots(pids)
+
+            # TODO: close launchers
+
+        except:
+            self._log.exception('finalization error')
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _pmgr_control_cb(self, topic, msg):
+
+        cmd = msg['cmd']
+        arg = msg['arg']
+
+        self._log.debug('launcher got %s', msg)
+
+        if cmd == 'kill_pilots':
```

Comment:
then the following should be updated as well?
https://github.com/radical-cybertools/radical.pilot/blob/7f1cfee2b7ebfc2a4b5522159e5739ac18908c85/src/radical/pilot/pilot_manager.py#L819-L822
should

Reply:
Thanks, done.

Reply:
Thanks, Done.